Add `Integrator` entry point and developper documentation (#2916)
qutip/qutip: doc/internals/solver/integrator.rst

.. _integrator_internal:

Integrator Architecture and Design
##################################

The numerical solution of differential equations is a cornerstone of QuTiP's
time-evolution solvers. Because different quantum systems present varying behaviours,
no single numerical algorithm is optimal for all scenarios.

> The :class:`Integrator` framework provides a unified, backend-agnostic interface
> that wraps external Ordinary Differential Equation (ODE) libraries. This design
> allows solvers to select routines from multiple distinct libraries through a single,
> interchangeable class.

Reviewer: ```suggestion
The :class:`Integrator` framework provides a unified, backend-agnostic interface
that wraps external or custom Ordinary Differential Equation (ODE) libraries. This design
allows solvers to select routines from multiple distinct libraries through a single,
interchangeable class.
```

This design has the advantage of allowing QuTiP to wrap external ODE packages
that lack native event-detection features, guaranteeing consistent behavior across all
ODE providers.

Configuration and State Resetting
=================================

Numerical routines depend on parameters (such as tolerance limits
``atol``/``rtol``, step limits ``nsteps``, etc.) that must be tuned to achieve
optimal efficiency. These parameters are configurable by the developer or end
user via the ``options`` attribute passed down to the integrator instance.
The ``Integrator.options`` attribute operates as a live dictionary containing
the configuration state.

Because used libraries maintain internal state for their ODE solvers,
modifying values within the ``options`` dictionary does not always take effect immediately.
Activating the newly assigned options requires an explicit call to :meth:`reset()`.
This tears down the active backend instance and restarts it with the new options
while preserving the active state and time.

Derivative Format
=================

While ordinary ODE interfaces strictly expect functions to represent the right-hand side (RHS),

> quantum equation in QuTiP can profit from alternative formats.

Reviewer: ```suggestion
While ordinary ODE interfaces strictly expect functions to represent the right-hand side (RHS),
quantum solvers in QuTiP can profit from alternative formats.
```

Every integrator exposes an :attr:`RHS_format` class attribute string that dictates
the type of RHS object required during initialization:

* **"callable"**: The integrator expects a function matching the signature
  ``rhs(t: float, state: Data) -> Data``. This is the default for most ODE methods.
  For specific Cython-backed ODE implementations, if the passed function is a method of
  :class:`QobjEvo`, it will bypass standard Python overhead and hook directly into the

>   Cython bindings of ``QobjEvo.matmul`` using a ``RHS`` object.

Reviewer: It would be helpful to add a sentence describing what the arguments and return value of `rhs` are. Perhaps:

The ``rhs`` function returns the derivative of the `state` at time `t`.

* **"matrix"**: The integrator accepts a complex matrix formatted as a
  :class:`qutip.core.data.Data` object, solving the linear matrix differential equation
  $\frac{dX}{dt} = \text{RHS} \times X$.
* **"solver"**: The integrator takes the parent :class:`Solver` instance that instantiated it
  and builds the RHS internally. This is reserved for specialized integration methods that
  deliberately blend the underlying quantum physics with the numerical method.

Class Attributes Reference
==========================

Every subclass inheriting from :class:`Integrator` must define the following class-level attributes:

.. attribute:: integrator_options

   A class-level dictionary defining the comprehensive set of configuration keys supported
   by the numerical backend along with their default values.
   Upon instantiation, this is copied to an instance-level ``options`` dictionary.

> .. attribute:: RHS_format

Reviewer: ```suggestion
.. attribute:: rhs_format
```
I would personally rename this to all lowercase. Uppercase is not typically used in Python attribute names.

   A string identifier defining the expected structural format of the Right-Hand Side (RHS)
   derivative function (e.g., ``"callable"``, ``"matrix"``, or ``"solver"``).

.. attribute:: name

   A descriptive, human-readable string representation of the algorithm
   (e.g., ``"SciPy zvode Adams solver"``) used primarily for populating metadata
   inside the final :class:`Result` container.

Creating New Integrators
========================

To implement a custom numerical backend, create a subclass derived from :class:`Integrator`.

> At a minimum, your class must override :meth:`integrate`, :meth:`set_state`, :meth:`get_state`,
> and either :meth:`__init__` or :meth:`_prepare`:
> 
> .. code-block:: python

Reviewer: ```suggestion

Let's build a simple linear step integrator that updates the current state by adding `dt * derivative(t, state)`. This integrator won't provide very accurate results, but it will nicely illustrate what's required.

.. code-block:: python
```

    from qutip.solver.integrator import Integrator

    class LinearStepODE(Integrator):
        RHS_format = "callable"
        name = "Simple Linear Step"
        method = "linear"

>         # Uses default empty dict if no specialized integrator_options are defined

Reviewer: ```suggestion
        # An integrator may also define `integrator_options` here. If the attribute is not set,
        # a default empty dict is assumed, indicating that their are not special options for
        # this integrator.
```

        def _prepare(self):
            # __init__ automatically prepares the local options dict,
            # stores the rhs in self.derivative, and initializes self._is_set to False.

>             # Apply option here!

Reviewer: ```suggestion
            # Apply any specialized options here!
```

            # This will be called from __init__ and reset.
            pass

        def set_state(self, t, state):
            self.t = t
            self.state = state
            self._is_set = True

        def integrate(self, t):
            # Basic Euler step example
            self.state = self.state + (t - self.t) * self.derivative(self.t, self.state)
            self.t = t

        def get_state(self):
            return self.t, self.state

The base :class:`Integrator` class automatically provides default implementations for
the :meth:`run` :meth:`reset` and :meth:`mcstep` methods
(though the fallback version of :meth:`mcstep` can be very inefficient).

To expose your custom integrator to a specific physical solver,
register it using the solver class's target method:

.. code-block:: python

    from qutip.solver import SESolver

    SESolver.add_integrator(LinearStepODE, "linear")

Once added, the method is immediately accessible via the high-level functional
wrappers by configuring the options map (e.g., ``options={"method": "linear"}``).

> Registering the class with the parent :class:`Solver` base class makes it globally available to all solvers except stochastic solver.

Reviewer: ```suggestion

Registering the class with the parent :class:`Solver` base class makes it globally available to all solvers except stochastic solver.
```
Just making this a separate paragraph.

The RHS class
-------------
To optimize performance, certain Cython-backed integrators reuse memory when
computing derivatives by supporting an in-place functional signature:

.. code-block:: python

    rhs(t: float, state: Data, out: Data) -> Data

In this paradigm, the operation should be interpreted as returning the computed
derivative added directly into the existing ``out`` object. The function is
permitted to reuse, modify, or overwrite the ``out`` container as needed.
This pattern eliminates the considerable overhead of frequent memory allocations
and deallocations inside tight numerical integration loops. The returned object

> does not need to the same as the ``out`` input if immutable or otherwise not reusable.

Reviewer: ```suggestion
and deallocations inside tight numerical integration loops. The returned object
does not need to the same as the ``out`` input if is immutable or otherwise not reusable.
```

Author: it's?

These specialized integrators are marked as ``RHS_format="callable"``.
Internally, they rely on the ``RHS`` helper class.
This class acts as an interface layer that safely wraps an ordinary,
out-of-place python derivative function and transforms it into one that behaves
correctly when subjected to in-place calls.

> Within the QuTiP's solver, physical derivatives are usually
> computed via :meth:`QobjEvo.matmul_data`. The integrators automatically detects
> that method and bind directly to the low-level Cython implementation with
> in-place support.

Reviewer: ```suggestion
Within the QuTiP's solver, physical derivatives are usually
computed via :meth:`QobjEvo.matmul_data`. The integrators automatically detect
that method and bind directly to the low-level Cython implementation with
in-place support.
```
